reconcile/saas_auto_promotions_manager/merge_request_manager/merge_request_manager.py

```diff
@@ -214,8 +215,24 @@ def _unbatch_failed_mrs(self) -> None:
         and close the old batched MR. By doing so, we ensure that unrelated MRs are not blocking each other.
         Unbatched MRs are marked and will never be batched again.
         """
-        # TODO: implemented in follow-up MR
-        pass
+        open_mrs_after_unbatching: list[OpenMergeRequest] = []
+        for mr in self._open_mrs:
```

Comment:
The code in `_unbatch_failed_mrs` has some duplicates and complex loop/condition logic, better to extract a method to keep collection operation and single item operation seperate. Like this:

```python
 def _unbatch_mr(self, mr: OpenMergeRequest) -> bool:
    if not mr.is_batchable:
        return False
    if not mr.failed_mr_check:
        return False
    self._vcs.close_app_interface_mr(...)
    ...
    return True
```

```python
def _unbatch_failed_mrs(self) -> None:
    self._open_mrs = [mr for mr in self._open_mrs if not self._unbatch_mr(mr)]
````
```

```

Reply:
Nice idea! But would like to keep this in a single function for now. I think the cognitive load of the function is acceptable.

Reply:
OK, but try not to use multiple `continue` in a large loop, it's like `GOTO`, hard to track logic flow. Since this logic is quite simple, maybe a positive filter is cleaner:

```python
        if mr.is_batchable and mr.failed_mr_check:
            self._vcs.close_app_interface_mr(
                mr.raw,
                "Closing this MR because it failed MR check and isn't marked un-batchable yet.",
            )
            # Remember these hashes as unbatchable
            content_hashes = mr.content_hashes.split(ITEM_SEPARATOR)
            self._unbatchable_hashes.update(content_hashes)
        else:
            open_mrs_after_unbatching.append(mr)
```

Reply:
Very good point. Fixed :+1: